### ISSUE TITLE
Scratchpad geometry problems.

### ISSUE THREAD

**Issue by Ergus** [i3/i3#3936]
## I'm submitting a…

<pre>
[x] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

Scratchpad size defined with relative units breaks when using multiple monitors.

My actual configuration is:
```
for_window [instance="__scratchpad"] move scratchpad, resize set 100 ppt 80 ppt, move position 0 ppt 0 ppt, move scratchpad
exec uxterm -name __scratchpad
```
Which seems to work more or less fine in a single monitor, but it is still a little hacky. So as it is a kind of hack I am not sure is this is a bug of the actual implementation or a missing feature to do it propperly.

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version:

Binary i3 version:  4.17.1 (2019-08-30) © 2009 Michael Stapelberg and contributors
(Getting version from running i3, press ctrl-c to abort…)
Running i3 version: 4.17.1 (2019-08-30) (pid 232938)
Loaded i3 config: /home/ergo/.config/i3/config (Last modified: Tue 11 Feb 2020 19:58:28 CET, 703 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
 
</pre>

<pre>
- Linux Distribution & Version: Arch Linux
- Are you using a compositor (e.g., xcompmgr or compton): No
</pre>

**Comment by Airblader** [comment]
Can you please elaborate on this? You said it "breaks", but you didn't really describe how it breaks: what do you expect and what do you see instead?

**Comment by Ergus** [comment]
You can see how it looks when opening from different desktops (it is the vertical "window" with the blue border)

![image](https://user-images.githubusercontent.com/6975176/74426975-9dcf8b80-4e56-11ea-9f15-8e5237977bcb.png)

![image](https://user-images.githubusercontent.com/6975176/74427138-ce172a00-4e56-11ea-8925-a7803bd0de1c.png)

The same with only 2 desktops in horizontal (so a simpler shape):

![image](https://user-images.githubusercontent.com/6975176/74427382-4a117200-4e57-11ea-9624-9565dfe1e827.png)

**Comment by orestisfl** [comment]
I can't reproduce this. Can you retry with 4.18 and upload a log file please?

https://i3wm.org/docs/debugging.html

**Comment by Ergus** [comment]
Hi, trying after the updates I get a better behavior. But something is still not working fine:

with this config:

```
# scratchpad with a terminal by default
for_window [instance="__scratchpad"] move scratchpad, resize set 100 ppt 80 ppt, move position 0 px 0 px, move scratchpad

# for_window [instance="__scratchpad"] move scratchpad
exec $term -name __scratchpad
bindsym F12 scratchpad show
```
I get

![image](https://user-images.githubusercontent.com/6975176/79031057-9d284c80-7b9c-11ea-9bfa-dbcc76adb37c.png)


As you can see the size is not 100 and 80% of the screen (in any of them). 

This is much better than what I had before, but maybe this must be fixed as well?

**Comment by orestisfl** [comment]
Yes, this needs fixing but I can't reproduce it. Please provide a log file.

**Comment by Ergus** [comment]
Enabling logging the probllem became the same as before. I added the log: https://logs.i3wm.org/logs/5714619301625856.bz2

**Comment by orestisfl** [comment]
Ok this is amazing

First of all, I think a workaround here is to remove the first `move scratchpad`. It is more correct this way because `resize set 100 80 ppt` will refer to the pseudo-output of the scratchpad pseudo-workspace, not your primary screen.

And this leads us to this:
```
11/04/20 17:32:28 - scratchpad.c:scratchpad_fix_resolution:259 - output eDP1's resolution: (1366, 312) 1366 x 768
11/04/20 17:32:28 - scratchpad.c:scratchpad_fix_resolution:259 - output DP2's resolution: (0, 312) 1366 x 768
11/04/20 17:32:28 - scratchpad.c:scratchpad_fix_resolution:259 - output HDMI1's resolution: (2732, 0) 1920 x 1080
11/04/20 17:32:28 - scratchpad.c:scratchpad_fix_resolution:273 - new width = 1311360, new height = 34560
```

Because of the unique combination of widths of your monitors, the LCM (used because of #632) explodes, leading to a huge scartchpad workspace.

**Comment by orestisfl** [comment]
@stapelberg you know the history here, do you think that `scratchpad_fix_resolution` is still relevant here? I have completely removed the function and the tests still pass.

**Comment by Ergus** [comment]
First of all; I think a workaround here is to remove the first `move scratchpad`. 

I did it
```
# scratchpad with a terminal by default
for_window [instance="__scratchpad"] resize set 100 ppt 80 ppt, move position 0 px 0 px, move scratchpad

# for_window [instance="__scratchpad"] move scratchpad
exec $term -name __scratchpad
bindsym F12 scratchpad show
```

 and then the `resize set 100 80 ppt` doesn't apply as you can see.

![image](https://user-images.githubusercontent.com/6975176/79055609-b7226780-7c4e-11ea-8d74-d1033ac22eab.png)

**Comment by stapelberg** [comment]
> @stapelberg you know the history here, do you think that `scratchpad_fix_resolution` is still relevant here? I have completely removed the function and the tests still pass.

I don’t remember any more details about the history, sorry.

I recommend we apply [Chesterton’s Fence](https://fs.blog/2020/03/chestertons-fence/) here: reproduce the original problem, then reason about why the original fix is no longer required. Maybe the code changed in a way that made it superflous? Maybe the test was broken to begin with, or broke over time?

**Comment by orestisfl** [comment]
I tried to run the testsuite but it got hanged and I didn't have the energy to debug 4.2's testsuite.

**Comment by stapelberg** [comment]
Did you try running just the 1 specific test case that was added? That way, unrelated testsuite issues can be ignored.

**Comment by orestisfl** [comment]
Ok! I was able to run it like this:

```diff
diff --git a/testcases/lib/StartXDummy.pm b/testcases/lib/StartXDummy.pm
index 5c739fca0..19eb9b6f3 100644
--- a/testcases/lib/StartXDummy.pm
+++ b/testcases/lib/StartXDummy.pm
@@ -93,8 +93,7 @@ sub start_xdummy {
         # We use -config /dev/null to prevent Xdummy from using the system
         # Xorg configuration. The tests should be independant from the
         # actual system X configuration.
-        my $socket = fork_xserver($displaynum, './Xdummy', ":$displaynum",
-                '-config', '/dev/null', '-nolisten', 'tcp');
+        my $socket = fork_xserver($displaynum, 'Xephyr', ":$displaynum", '-screen', '1280x800', '-nolisten', 'tcp');
         push(@displays, ":$displaynum");
         push(@sockets_waiting, $socket);
         $displaynum++;
```

**Comment by orestisfl** [comment]
Ok the test does not fail on 3cdc5c5369efb10cdd46498cf478e824fb1a870f with the testcase from 6ba09444308ac4c8c55beaef416cbaafe7470fb0 :/

**Comment by orestisfl** [comment]
But I was able to reproduce the issue with `fake-outputs 400x250+0+0,333x333+400+0` in `next` if I remove `scratchpad_fix_resolution`. So we have something to work with without needing to jump to the past.
